io/zenoh-transport/src/unicast/shm/transport.rs

```diff
@@ -0,0 +1,329 @@
+        // check if the link is ours
+        if self.zlinkget_complementary(link).await.is_err() {
+            bail!(
+                "Can not delete Link {} with peer: {}",
+                link,
+                self.config.zid
+            )
+        }
+
+        // Notify the callback
+        if let Some(callback) = zread!(self.callback).as_ref() {
+            callback.del_link(Link::from(link));
+        }
+
+        self.delete().await
+    }
+}
+
+#[async_trait]
+impl TransportUnicastInnerTrait for ShmTransportUnicastInner {
+    /*************************************/
+    /*            ACCESSORS              */
+    /*************************************/
+    fn set_callback(&self, callback: Arc<dyn TransportPeerEventHandler>) {
+        let mut guard = zwrite!(self.callback);
+        *guard = Some(callback);
+    }
+
+    async fn get_alive(&self) -> AsyncMutexGuard<'_, bool> {
+        zasynclock!(self.alive)
+    }
+
+    fn get_links(&self) -> Vec<LinkUnicast> {
+        let guard = async_std::task::block_on(async { zasyncread!(self.link) });
+        [guard.clone()].to_vec()
+    }
+
+    fn get_zid(&self) -> ZenohId {
+        self.config.zid
+    }
+
+    fn get_whatami(&self) -> WhatAmI {
+        self.config.whatami
+    }
+
+    #[cfg(feature = "shared-memory")]
+    fn is_shm(&self) -> bool {
+        self.config.is_shm
+    }
+
+    fn is_qos(&self) -> bool {
+        self.config.is_qos
+    }
+
+    fn get_callback(&self) -> Option<Arc<dyn TransportPeerEventHandler>> {
+        zread!(self.callback).clone()
+    }
+
+    fn get_config(&self) -> &TransportConfigUnicast {
+        &self.config
+    }
+
+    /*************************************/
+    /*                TX                 */
+    /*************************************/
+    fn schedule(&self, msg: NetworkMessage) -> ZResult<()> {
+        self.internal_schedule(msg)
+    }
+
+    fn start_tx(
+        &self,
+        _link: &LinkUnicast,
+        executor: &TransportExecutor,
+        keep_alive: Duration,
+        _batch_size: u16,
+    ) -> ZResult<()> {
+        self.start_keepalive(executor, keep_alive);
+        Ok(())
+    }
+
+    fn start_rx(&self, _link: &LinkUnicast, lease: Duration, batch_size: u16) -> ZResult<()> {
+        self.internal_start_rx(lease, batch_size);
+        Ok(())
+    }
+
+    /*************************************/
+    /*               LINK                */
+    /*************************************/
+    fn add_link(&self, link: LinkUnicast, _direction: LinkUnicastDirection) -> ZResult<()> {
+        log::trace!("Adding link: {}", link);
+
+        #[cfg(not(feature = "transport_shm"))]
+        bail!(
+            "Can not add Link {} with peer {}: link already exists and only unique link is supported!",
+            link,
+            self.config.zid,
+        );
+
+        #[cfg(feature = "transport_shm")]
+        async_std::task::block_on(async {
```

Comment:
If required, why not changing the `add_link` and `del_link` internal signature to `async`?